service/controller/resource/clusterupgrade/create.go

```diff
@@ -111,57 +112,70 @@ func (r *Resource) ensureMasterHasUpgraded(ctx context.Context, cluster capiv1al
 }
 
 func (r *Resource) propagateReleaseToMachinePools(ctx context.Context, cr capiv1alpha3.Cluster) error {
-	lst := expcapiv1alpha3.MachinePoolList{}
-	err := r.ctrlClient.List(ctx, &lst, client.MatchingLabels{capiv1alpha3.ClusterLabelName: key.ClusterName(&cr)})
+	lst, err := helpers.GetMachinePoolsByMetadata(ctx, r.ctrlClient, cr.ObjectMeta)
 	if err != nil {
 		return microerror.Mask(err)
 	}
 
-	if !hasMachinePoolsUpgradeCompleted(lst.Items, cr.Labels[label.ReleaseVersion]) {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "machinepool still upgrading; skipping release propagation")
-		return nil
-	}
+	allStartedNodePoolUpgradesAreCompleted := true
+	var machinePoolUpgradeCandidate *capiv1alpha3exp.MachinePool
 
 	for _, mp := range lst.Items {
-		if cr.Labels[label.ReleaseVersion] != mp.Labels[label.ReleaseVersion] ||
-			cr.Labels[label.AzureOperatorVersion] != mp.Labels[label.AzureOperatorVersion] {
-
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating release to machinepool %q", mp.Name))
-
-			mp.Labels[label.ReleaseVersion] = cr.Labels[label.ReleaseVersion]
-			mp.Labels[label.AzureOperatorVersion] = cr.Labels[label.AzureOperatorVersion]
-
-			err = r.ctrlClient.Update(ctx, &mp)
-			if apierrors.IsConflict(err) {
-				r.logger.LogCtx(ctx, "level", "debug", "message", "conflict trying to save object in k8s API concurrently", "stack", microerror.JSON(microerror.Mask(err)))
-				r.logger.LogCtx(ctx, "level", "debug", "message", "cancelling resource")
-				return nil
-			} else if err != nil {
+		machinePool := mp
+
+		if cr.Labels[label.ReleaseVersion] == machinePool.Labels[label.ReleaseVersion] &&
+			cr.Labels[label.AzureOperatorVersion] == machinePool.Labels[label.AzureOperatorVersion] {
+			// Upgrade was initiated for this node pool, let's check if it has
+			// been completed.
+			nodePoolUpgradeInProgressOrPending, err := upgrade.IsNodePoolUpgradeInProgressOrPending(
+				ctx,
+				r.ctrlClient,
+				&machinePool,
+				cr.Labels[label.ReleaseVersion],
+				cr.Labels[label.AzureOperatorVersion])
+			if err != nil {
 				return microerror.Mask(err)
 			}
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated release to machinepool %q", mp.Name))
+			if nodePoolUpgradeInProgressOrPending {
+				// This node pool is still being upgraded, so we want to wait
+				// for this upgrade to be completed.
+				allStartedNodePoolUpgradesAreCompleted = false
+				break
+			}
+		}
+
+		if machinePoolUpgradeCandidate == nil &&
+			(cr.Labels[label.ReleaseVersion] != machinePool.Labels[label.ReleaseVersion] ||
+				cr.Labels[label.AzureOperatorVersion] != machinePool.Labels[label.AzureOperatorVersion]) {
+
+			// We did not already find next node pool to be upgraded, and this
+			// one does not have version labels updated, so this node pool will
+			// be upgraded next.
+			machinePoolUpgradeCandidate = &machinePool
```

Comment:
Here we pick the node pool candidate for the next upgrade. (so a potential upgrade, since we still iterate through the rest of the nodes to check if some of them is not done upgrading)

Reply:
While reviewing this function, I'd prefer it to be split into smaller components. One possibility could be with following three functions:
- One that checks if any of the present `MachinePools` is upgrading.
  - If yes, then we can just return at that point since we don't want to start next one.
- One that filters out already upgraded node pools from `MachinePoolList`.
  - In case `len(leftovers) == 0` => Upgrade is done.
  - Otherwise one can pick `1:n` elements from `leftovers` for next upgrade.
- One that ensures given `MachinePool` has up-to-date labels.

WDYT?

Reply:
Yeah, that sounds good.